### PR TITLE
Add new command `pub outdated`

### PR DESCRIPTION
**Pull request by sigurdm**
This command intends to make it easier to get an overview of your dependencies.

Support for marking null-safe dependencies is not included in this PR.

The files test/outdated/goldens/ contains test sessions that demonstrates the output of the command with different options (though without color).

### CODE REVIEW

File: lib/src/log.dart [hunk 1]
```diff
@@ -380,6 +380,18 @@ Future<T> progress<T>(String message, Future<T> Function() callback,
   return callback().whenComplete(progress.stop);
 }
 
+/// Like [progress] but erases the message once done.
+Future<T> spinner<T>(String message, Future<T> Function() callback,
+    {bool fine = false}) {
```
Comment (jonasfj) [inline lib/src/log.dart]: `fine` is never used :)

File: lib/src/command/outdated.dart [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    log.message('Found no outdated packages');
+    return;
+  }
+
+  final formattedRows = <FormattedRow>[
+    FormattedRow(
+        null,
+        ['Package', 'Current', 'Upgradable', 'Resolvable', 'Latest']
+            .map((s) => _FormattedString(s, format: log.bold))
+            .toList()),
```
Comment (jonasfj) [inline lib/src/command/outdated.dart]: I would suggest adding:
```dart
...rows.where((r) => r.kind != _DependencyKind.production).map(marker.formatRow)
FormattedRow(['dev_dependencies', ]),
...rows.where((r) => r.kind != _DependencyKind.dev).map(marker.formatRow)
```

Or something like this... The ` if (lastKind != row.kind) {` is a bit sketchy :)

File: lib/src/command/outdated.dart [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+class FormattedRow {
+  final _PackageDetails _row;
+  final List<_FormattedString> columns;
+  FormattedRow(this._row, this.columns);
+  _DependencyKind get kind => _row?.kind;
+}
+
+abstract class _Marker {
+  Future<FormattedRow> formatRow(_PackageDetails packageDetails);
```
Comment (jonasfj) [inline lib/src/command/outdated.dart]: Avoid classes when functions types can do...

File: lib/src/command/outdated.dart (hunk 2 above)
Comment (jonasfj) [reply lib/src/command/outdated.dart]: You can still do this an async manner..

File: lib/src/command/outdated.dart [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+      entrypoint.root.name,
+      ...immediateDependencies.map((d) => d.name)
+    };
+    for (final id in [
+      if (includeDevDependencies) ...entrypoint.lockFile.packages.values,
+      ...unconstrainedSolveResult.packages
+    ]) {
+      final name = id.name;
+      if (visited.contains(name)) continue;
+      visited.add(name);
```
Comment (jonasfj) [inline lib/src/command/outdated.dart]: just do: ` if (!visited.add(name)) continue;`

File: lib/src/command/outdated.dart [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+        allowed: ['outdated', 'none'],
+        defaultsTo: 'outdated');
+  }
+
+  @override
+  Future run() async {
+    final includeDevDependencies = argResults['dev-dependencies'];
+
+    final oldVerbosity = log.verbosity;
+    // Unless the user overrides the verbosity, we want to filter out the
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/command/outdated.dart]: Great idea!
Done

File: lib/src/command/outdated.dart [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+    argParser.addMultiOption('include',
+        allowed: ['pre-releases', 'transitive', 'up-to-date'],
+        help:
+            'pre-releases: include pre-releases when reporting latest version.\n'
+            'up-to-date: include dependencies that are already at the latest version.\n'
+            'transitive: include transitive dependencies in the reports',
+        valueHelp: 'optionA,optionB');
+
+    argParser.addMultiOption('exclude',
+        allowed: ['dev-dependencies'],
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/command/outdated.dart]: I reorganized the options

File: test/outdated/goldens/newer_versions.txt [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+transitive dependencies
+transitive   *1.2.3   *1.3.0      *1.3.0      2.0.0   
+transitive2  -        -           1.0.0       1.0.0   
+transitive3  -        -           1.0.0       1.0.0   
+
+3 dependencies are currently not `pubspec.lock`'ed at the `Upgradable` version.
+They can be upgraded with `pub upgrade`.
+
+4 dependencies can be upgraded to the ‘Resolvable’ version by updating the
+constraints in `pubspec.yaml`.
```
Comment (sigurdm) [inline test/outdated/goldens/newer_versions.txt]: @kwalrath can you help with the wording of these sentences?

Comment (kwalrath) [reply test/outdated/goldens/newer_versions.txt]: How about something like this:

```
3 upgradable dependencies are locked (in pubspec.lock) to older versions.
To update these dependencies, use `pub upgrade`.

4 dependencies are constrained to versions that are older than a resolvable version.
To update these dependencies, edit pubspec.yaml.
```

Comment (sigurdm) [reply test/outdated/goldens/newer_versions.txt]: That's nice. Thanks!

**Comment by sigurdm** [comment]
This command intends to make it easier to get an overview of your dependencies.

Support for marking null-safe dependencies is not included in this commit.

**Pull request merged**